File: Max Heap/max.cpp
#include <iostream>
#include <vector>
using namespace std;

void swap(int *u, int *v)
{
  int temp = *v;
  *v = *u;
  *u = temp;
}
void max(vector<int> &hT, int i)
{
  int size = hT.size();
  int largest = i;
  int l = 2 * i + 1;
  int r = 2 * i + 2;
  if (l < size && hT[l] > hT[largest])
    largest = l;
  if (r < size && hT[r] > hT[largest])
    largest = r;

  if (largest != i)
  {
    swap(&hT[i], &hT[largest]);
    max(hT, largest);
  }
}
void insert(vector<int> &hT, int newNum)
{
  int size = hT.size();
  if (size == 0)
  {
    hT.push_back(newNum);
  }
  else
  {
    hT.push_back(newNum);
    for (int i = size / 2 - 1; i >= 0; i--)
    {
      max(hT, i);
    }
  }
}
void deleteNode(vector<int> &hT, int num)
{
  int size = hT.size();
  int i;
  for (i = 0; i < size; i++)
  {
    if (num == hT[i])
      break;
  }
  swap(&hT[i], &hT[size - 1]);

  hT.pop_back();
  for (int i = size / 2 - 1; i >= 0; i--)
  {
    max(hT, i);
  }
}
void print(vector<int> &hT)
{
  for (int i = 0; i < hT.size(); ++i)
    cout << hT[i] << " ";
  cout << "\n";
}

int main()
{
  vector<int> heapTree;

  insert(heapTree, 7);
  insert(heapTree, 9);
  insert(heapTree, 6);
  insert(heapTree, 5);
  insert(heapTree, 2);

  cout << "Max-Heap array: ";
 
  print(heapTree);

  deleteNode(heapTree, 6);

  cout << "After deleting an element: ";

  print(heapTree);
}

File: Max Heap/max.py
def max(arr, j, i):
    largest = i
    l = 2 * i + 1
    r = 2 * i + 2 
    
    if l < j and arr[i] < arr[l]:
        largest = l
    
    if r < j and arr[largest] < arr[r]:
        largest = r
    
    if largest != i:
        arr[i],arr[largest] = arr[largest],arr[i]
        max(arr, j, largest)

def insert(array, Num):
    size = len(array)
    if size == 0:
        array.append(Num)
    else:
        array.append(Num)
        for i in range((size//2)-1, -1, -1):
            max(array, size, i)

def deleteNode(array, num):
    size = len(array)
    i = 0
    for i in range(0, size):
        if num == array[i]:
            break
        
    array[i], array[size-1] = array[size-1], array[i]

    array.remove(num)
    
    for i in range((len(array)//2)-1, -1, -1):
        max(array, len(array), i)
    
arr = []

insert(arr, 17)
insert(arr, 15)
insert(arr, 13)
insert(arr, 16)
insert(arr, 12)

print ("Max-Heap array: " + str(arr))

deleteNode(arr, 15)
print("After deleting an element: " + str(arr))